source/Octopus.Versioning/Maven/MavenVersion.cs

```diff
@@ -78,15 +79,7 @@ public override bool Equals(object obj)
 
         public override int GetHashCode()
         {
-            unchecked
-            {
-                var hashCode = Major;
-                hashCode = (hashCode * 397) ^ Minor;
-                hashCode = (hashCode * 397) ^ Patch;
-                hashCode = (hashCode * 397) ^ Revision;
-                hashCode = (hashCode * 397) ^ (ReleaseLabels != null ? ReleaseLabels.GetHashCode() : 0);
-                return (int)hashCode;
-            }
+            return VersionComparer.Default.GetHashCode(this);
```

Comment:
I don't think the default `VersionComparer` will include `ReleaseLabels` in the comparison. How will this change impact the intended comparison behaviour?

Reply:
From [the comment in VersionComparison](https://github.com/OctopusDeploy/Versioning/blob/main/source/Octopus.Versioning/Semver/VersionComparison.cs) the Default compares all parts of the version. Though this _is_ in the Semver directory so I'm not certain how it applies to Docker/Maven versions.

Reply:
And [here](https://github.com/OctopusDeploy/Versioning/blob/main/source/Octopus.Versioning/Semver/VersionComparer.cs#L74) is the `GetHashCode` implementation - the default checks on version + release label + metadata. I've moved `VersionComparer` and `VersionComparison` out of the Semver directory since it's not something specific to Semver, and we can use it for the other versions (at least this is what I'm thinking).

Reply:
Oh 🤔 yeah that doesn't look right

Reply:
After having a look at this again, I think the `VersionComparer` is kind of a semver specific thing. Semver pre-releases can be anything from the looks of it, whereas Maven prereleases can only be something like `SNAPSHOT-blah` or `ALPHA-blah`. So the default hashcode will skip over these kinds of Maven versions since `IsPrerelease` will return false for a Maven version with `bar` as a release label. So we can't really rely on the default hashcode.

I've reverted the Maven hash code, and updated it so that it gets the hash code for each release label (my guess is that it was failing before as it was taking the hash of the entire enumerable at once, which sounds like something that can change). I've also added `GetHashCode` for `DockerTag` similar to the Maven one.